In [23]:
# Import all required dependencies

import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

from sqlalchemy import create_engine
import psycopg2
from config import db_password


### Import the all_tables_cleaned.csv

(from Sean Tyson)

In [3]:
# file directory
csv_dir = "C:/Users/mamoo/UT_DataScience/UTA-VIRT-DATA-PT-02-2021-U-B/FinalProject/Resources/Clean_csv/"

# import the all_player.csv
all_tables_df = pd.read_csv(f"{csv_dir}all_tables_cleaned.csv", encoding= 'unicode_escape')
pd.set_option("display.max_columns", None) # To show all columns explicitly
all_tables_df

Pos                  Team  Year  Games_Played  Wins  Losses  Draws  \
0      1         Columbus Crew  2004            30    12       5     13   
1      2  Sporting Kansas City  2004            30    14       9      7   
2      3             LA Galaxy  2004            30    11       9     10   
3      4           D.C. United  2004            30    11      10      9   
4      5       Colorado Rapids  2004            30    10       9     11   
..   ...                   ...   ...           ...   ...     ...    ...   
324   23   Minnesota United FC  2021             7     2       4      1   
325   24             FC Dallas  2021             7     1       3      3   
326   25            Toronto FC  2021             7     1       4      2   
327   26       Chicago Fire FC  2021             7     1       5      1   
328   27         FC Cincinnati  2021             6     1       4      1   

     Goals_For  Goals_Against  Goal_Differential  Points Conference  
0           40             32                  8      49    Overall  
1           38             30                  8      49    Overall  
2           42             40                  2      43    Overall  
3           43             42                  1      42    Overall  
4           29             32                 -3      41    Overall  
..         ...            ...                ...     ...        ...  
324          6             11                 -5       7    Overall  
325          8             11                 -3       6    Overall  
326          8             12                 -4       5    Overall  
327          4             11                 -7       4    Overall  
328          6             15                 -9       4    Overall  

[329 rows x 12 columns]

In [5]:
# Check all the column headers.
all_tables_df.columns

Index(['Pos', 'Team', 'Year', 'Games_Played', 'Wins', 'Losses', 'Draws',
       'Goals_For', 'Goals_Against', 'Goal_Differential', 'Points',
       'Conference'],
      dtype='object')

In [14]:
# Change the 'Team' header to 'Club'
all_tables_df = all_tables_df.rename(columns={"Team": "Club"}) 
all_tables_df.head(2)

Pos                  Club  Year  Games_Played  Wins  Losses  Draws  \
0    1         Columbus Crew  2004            30    12       5     13   
1    2  Sporting Kansas City  2004            30    14       9      7   

   Goals_For  Goals_Against  Goal_Differential  Points Conference  
0         40             32                  8      49    Overall  
1         38             30                  8      49    Overall

In [15]:
all_tables_df.isnull().sum()

Pos                  0
Club                 0
Year                 0
Games_Played         0
Wins                 0
Losses               0
Draws                0
Goals_For            0
Goals_Against        0
Goal_Differential    0
Points               0
Conference           0
dtype: int64

In [16]:
all_tables_df.nunique()

Pos                  27
Club                 28
Year                 18
Games_Played         10
Wins                 22
Losses               24
Draws                18
Goals_For            63
Goals_Against        63
Goal_Differential    65
Points               65
Conference            1
dtype: int64

In [17]:
all_tables_df.describe()

Pos         Year  Games_Played        Wins      Losses  \
count  329.000000   329.000000    329.000000  329.000000  329.000000   
mean    10.279635  2013.860182     29.920973   11.100304   11.100304   
std      6.270885     5.026627      7.513797    4.254728    4.517476   
min      1.000000  2004.000000      6.000000    1.000000    0.000000   
25%      5.000000  2010.000000     30.000000    8.000000    8.000000   
50%     10.000000  2014.000000     34.000000   11.000000   11.000000   
75%     15.000000  2018.000000     34.000000   14.000000   14.000000   
max     27.000000  2021.000000     34.000000   22.000000   24.000000   

            Draws   Goals_For  Goals_Against  Goal_Differential      Points  
count  329.000000  329.000000     329.000000         329.000000  329.000000  
mean     7.720365   41.647416      41.647416           0.000000   41.021277  
std      3.262067   13.868184      13.976145          13.054996   13.900216  
min      0.000000    4.000000       3.000000         -44.000000    4.000000  
25%      6.000000   36.000000      36.000000          -7.000000   34.000000  
50%      8.000000   43.000000      43.000000           1.000000   42.000000  
75%     10.000000   50.000000      50.000000           8.000000   51.000000  
max     18.000000   85.000000      75.000000          48.000000   72.000000

In [18]:
all_tables_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 329 entries, 0 to 328
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Pos                329 non-null    int64 
 1   Club               329 non-null    object
 2   Year               329 non-null    int64 
 3   Games_Played       329 non-null    int64 
 4   Wins               329 non-null    int64 
 5   Losses             329 non-null    int64 
 6   Draws              329 non-null    int64 
 7   Goals_For          329 non-null    int64 
 8   Goals_Against      329 non-null    int64 
 9   Goal_Differential  329 non-null    int64 
 10  Points             329 non-null    int64 
 11  Conference         329 non-null    object
dtypes: int64(10), object(2)
memory usage: 31.0+ KB


The **all_tables_df** is now **cleaned** with total of **12 columns, 329 rows** of non-null data with consistent data types. 

### Export the clean df to .csv

In [19]:
# Export the clean dataframe to "all_players_clean.csv"
csv_dir = "C:/Users/mamoo/UT_DataScience/UTA-VIRT-DATA-PT-02-2021-U-B/FinalProject/Resources/Clean_csv/"
all_tables_df.to_csv(f"{csv_dir}all_tables_cleaned.csv")

In [20]:
# Group by "Club" and "Year"
all_tables_grouped = all_tables_df.groupby(['Club','Year']).sum()
all_tables_grouped

Pos  Games_Played  Wins  Losses  Draws  \
Club                   Year                                           
Atlanta United FC      2017    4            34    15       9     10   
                       2018    2            34    21       7      6   
                       2019    3            34    18      12      4   
                       2020   23            23     6      13      4   
                       2021   13             7     2       1      4   
...                          ...           ...   ...     ...    ...   
Vancouver Whitecaps FC 2017    8            34    15      12      7   
                       2018   14            34    13      13      8   
                       2019   23            34     8      16     10   
                       2020   17            23     9      14      0   
                       2021   21             7     2       4      1   

                             Goals_For  Goals_Against  Goal_Differential  \
Club                   Year                                                
Atlanta United FC      2017         70             40                 30   
                       2018         70             44                 26   
                       2019         58             43                 15   
                       2020         23             30                 -7   
                       2021          9              7                  2   
...                                ...            ...                ...   
Vancouver Whitecaps FC 2017         50             49                  1   
                       2018         54             67                -13   
                       2019         37             59                -22   
                       2020         27             44                -17   
                       2021          6              9                 -3   

                             Points  
Club                   Year          
Atlanta United FC      2017      55  
                       2018      69  
                       2019      58  
                       2020      22  
                       2021      10  
...                             ...  
Vancouver Whitecaps FC 2017      52  
                       2018      47  
                       2019      34  
                       2020      27  
                       2021       7  

[329 rows x 9 columns]

In [21]:
# Export the grouped dataframe to "all_players_grouped.csv"
csv_dir = "C:/Users/mamoo/UT_DataScience/UTA-VIRT-DATA-PT-02-2021-U-B/FinalProject/Resources/Clean_csv/"
all_tables_grouped.to_csv(f"{csv_dir}all_tables_grouped.csv")

## Export the DataFrame(s) to PostgreSQL Database

In [24]:
#Add the code to create the connection to the PostgreSQL database.
# For our local server, the connection string will be as follows:
db_link = f"postgres://postgres:{db_password}@127.0.0.1:5432/major_league_soccer"    

In [25]:
# Create the database engine 
engine = create_engine(db_link)

In [28]:
# add the all_tables_cleaned DataFrame to a SQL database.
all_tables_df.to_sql(name='mls_tables', con=engine, if_exists='replace')

In [29]:
# add the all_tables_grouped DataFrame to a SQL database.
all_tables_grouped.to_sql(name='tables_grouped', con=engine, if_exists='replace')